### ISSUE TITLE
"array: true" plugin parameter throws error when jsPsych.timelineVariable() is used (audio-button-response plugin)

### ISSUE THREAD

**Issue by devsda1** [jspsych/jsPsych#3412]
While updating a task programmed in jsPsych 7.x to 8.x, I found that supplying the option "choices: jsPsych.timelineVariable('choices')" to the audio-button-response plugin now causes the following error:

A non-array value (`function() { return fnGetImmediate(getArg); }`) was provided for the array parameter "choices" in the "audio-button-response" plugin. Please make sure that "choices" is an array.

The cause of this seems to be the newly strict handling of "array: true" for plugin parameters. It's true - the value here will be a function, not an array. 

But surely when a plugin sets "array: true", it should not break the use of timeline variables with that parameter?

**Comment by bjoluc** [comment]
Yup, fully agree – thanks @devsda1! I'll look into this on Thursday.

**Comment by bjoluc** [comment]
@devsda1 Timeline variable evaluation for array parameters works fine; your error message suggests that your timeline variable evaluates to a function instead of an array. Can you take a look at it again?

**Comment by jodeleeuw** [comment]
@devsda1 Just checking in to see if the source of the problem got tracked down?

**Comment by devsda1** [comment]
@jodeleeuw @bjoluc Sorry for the delay. I looked at this again. The code that worked in jsPsych 6.x-7.x provided a "choices" parameter for audio-button-response with the value jsPsych.timelineVariable('choices'), and the 'choices' timeline variable was in fact a function that returned an array value. So in effect, there were two "nested" functions. This is necessary for the task because the participant's language may not be known until after the timeline starts running (i.e. after an initial selection using html-button-response). We use functions fairly ubiquitously in our tasks in order to select plugin parameter values according to language. 

Based on your reply that "Timeline variable evaluation for array parameters works fine", I reasoned that my problem might be that the parameter value is a function that returns a function, and this "unwrapping" of function values no longer happens. So I tried supplying a function that returns an array (as I assume jsPsych.timelineVariable would do), and I get the same problem.

Since the ability to supply a function for a plugin parameter in order to defer evaluation of the value is quite fundamental to jsPsych, I'm confused about why jsPsych 8.x is throwing an error when it receives a function value. Shouldn't it call the function (and ideally, keep calling function values until it receives a non-function), and only then thrown an error if a type constraint (e.g. array) on the resulting value is not satisfied?

**Comment by devsda1** [comment]
@jodeleeuw @bjoluc My code that worked in 6.x/7.x populated certain timeline variables with function values in order to defer evaluation until the trial is about to run. The reason for that is to provide the correct content for the participant's language, which is selected only after the outer timeline starts running. 

In 6.x/7.x, jsPsych helpfully called functions that were returned by timelineVariable() (i.e. by the function provided by timelineVariable). 

In 8.x, it does not. So in this situation, rather than e.g. simply providing jsPsych.timelineVariable('choices') as the value of the 'choices' parameter, I need to do something like this:

    function resolveFunctions(val) {
        if (typeof(val) === 'function') {
            return resolveFunctions(val());
        }
        if (typeof(val) === 'object') {
            // works for objects and arrays
            for (let name in val) {
                val[name] = resolveFunctions(val[name]);
            }
        }
        return val;
    }

    function resolveTimelineVarLater(name) {
        return function() {
            return resolveFunctions(jsPsych.timelineVariable(name));
        }
    }

    // ...

    timeline.push({
        // ...
        choices: resolveTimelineVarLater('choices'),
        // ...
    });

I'm not sure that a value obtained by jsPsych from a timeline variable needs to be treated differently from one supplied as the parameter value itself, in terms of how functions are resolved to values and (if the plugin specifies a type constraint) the value types are ultimately checked. I didn't expect that to be the case, anyway. But if I need to have extra code like the above in my task to handle this, at least now I know how to do it.

**Comment by devsda1** [comment]
@bjoluc Thanks for the information! Not sure if I read the code correctly, but it looks like the new version also will only do one step of calling a function, whereas in 7.x replaceFunctionsWithValues is used, which recursively walks an object or array in a similar (but more jsPsych specific) way as my "resolveFunctions" code above? 

A maximally magic/flexible approach could be for replaceFunctionsWithValues (or something similar) to also resolve timeline variables if they are encountered during the walk. Then, whatever is supplied as a parameter value, any contained functions or timeline variables will be replaced until all values are resolved. (However if you need to resolve timeline variables from a parent timeline, which I haven't used myself but seems to be supported, I can see how it would become more complex to implement; I guess you'd have to inject the parent timelines into any recursion?)

**Comment by bjoluc** [comment]
> Not sure if I read the code correctly, but it looks like the new version also will only do one step of calling a function, whereas in 7.x replaceFunctionsWithValues is used, which recursively walks an object or array in a similar (but more jsPsych specific) way as my "resolveFunctions" code above?

The method is called recursively according to the plugin's parameter definitions. This means it no longer does any magic on recursive properties that are not listed in the plugin's parameter description (intended and noted in the v8 release notes) – something which you might previously have relied on as well.

> A maximally magic/flexible approach

I'm afraid, we have decided against the full magic version on purpose to avoid calling functions when it is not intended (e.g., on third-party objects).

**Comment by devsda1** [comment]
Ah I see, thanks. If I wanted to roll my own magic, I think I could just add a conditional evaluateTimelineVariable() in my recursive function. Is the TimelineVariable class visible from the context of the script that is calling jsPsych? I was thinking of something like "if (val instanceof TimelineVariable) {...}".
